### PR TITLE
[MVP-001] add typeScript. Enable the "no any" rule. Add start, build …

### PR DESCRIPTION
**Pull request by Okiana**
- [x] We must use the latest version of TypeScript.  

- [x]  TypeScript configuration must be as strict as possible, but we should not enforce a no any rule.

- [x] There should be a build script in the npm scripts.

- [x] Use Nodemon and create a dev watch mode.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "interview-me-backend",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
```
Comment (Teodor92) [inline package.json]: `index.js` is not going to be our main file - it's going to be `index.ts`

File: package.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  "scripts": {
+    "start": "npm run build:live",
+    "build": "tsc -p .",
+    "build:live": "nodemon --watch 'src/**/*.ts' --exec 'ts-node' src/index.ts",
```
Comment (Teodor92) [inline package.json]: Naming recommendation - `build:watch` it's a bit more clear than `live` :)

Comment (Teodor92) [reply package.json]: Also can you extract the config here to a `nodemon.json` file?

File: package.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "interview-me-backend",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "start": "npm run build:live",
+    "build": "tsc -p .",
```
Comment (Teodor92) [inline package.json]: No need to pass the `-p` flag - the tsc will get the tsconfig from the root

File: package.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    "build:live": "nodemon --watch 'src/**/*.ts' --exec 'ts-node' src/index.ts",
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (Teodor92) [inline package.json]: Lest delete scripts we currently do not use :)

File: package.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/QuestionQuest/interview-me-backend/issues"
+  },
+  "homepage": "https://github.com/QuestionQuest/interview-me-backend#readme",
+  "devDependencies": {
+    "@types/node": "^10.12.0",
```
Comment (Teodor92) [inline package.json]: Lets freze all project versions in order to prevent any unexpected behavior.

File: tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,59 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
+    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    // "lib": [],                             /* Specify library files to be included in the compilation. */
+    // "allowJs": true,                       /* Allow javascript files to be compiled. */
```
Comment (Teodor92) [inline tsconfig.json]: Lets switch this to false.

File: tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    // "checkJs": true,                       /* Report errors in .js files. */
+    // "jsx": "preserve",                     /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
+    // "declaration": true,                   /* Generates corresponding '.d.ts' file. */
+    // "declarationMap": true,                /* Generates a sourcemap for each corresponding '.d.ts' file. */
+    // "sourceMap": true,                     /* Generates corresponding '.map' file. */
```
Comment (Teodor92) [inline tsconfig.json]: We will need source maps in order to debug our app

File: tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    // "outFile": "./",                       /* Concatenate and emit output to single file. */
+    // "outDir": "./",                        /* Redirect output structure to the directory. */
```
Comment (Teodor92) [inline tsconfig.json]: Lets be explisit about the outDir. `dist` maybe?

File: tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    // "rootDir": "./",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
+    // "composite": true,                     /* Enable project compilation */
+    // "removeComments": true,                /* Do not emit comments to output. */
+    // "noEmit": true,                        /* Do not emit outputs. */
+    // "importHelpers": true,                 /* Import emit helpers from 'tslib'. */
```
Comment (Teodor92) [inline tsconfig.json]: I recommend enabling this

File: tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    "strict": true,                           /* Enable all strict type-checking options. */
+    "noImplicitAny": true,                 /* Raise error on expressions and declarations with an implied 'any' type. */
+    // "strictNullChecks": true,              /* Enable strict null checks. */
+    // "strictFunctionTypes": true,           /* Enable strict checking of function types. */
+    // "strictPropertyInitialization": true,  /* Enable strict checking of property initialization in classes. */
+    // "noImplicitThis": true,                /* Raise error on 'this' expressions with an implied 'any' type. */
+    // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */
+
+    /* Additional Checks */
+    // "noUnusedLocals": true,                /* Report errors on unused locals. */
```
Comment (Teodor92) [inline tsconfig.json]: Lets activate this - unused vars are not recommended.

File: tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    // "noUnusedParameters": true,            /* Report errors on unused parameters. */
```
Comment (Teodor92) [inline tsconfig.json]: Same as above

File: tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
```
Comment (Teodor92) [inline tsconfig.json]: It's always nice to be explicit about the code you are writing.

File: tsconfig.json [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */
```
Comment (Teodor92) [inline tsconfig.json]: It's a not a good idea to have falltrhough switches

File: tsconfig.json [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+    /* Module Resolution Options */
+    // "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
```
Comment (Teodor92) [inline tsconfig.json]: It's a good idea to activate this.

File: tsconfig.json [hunk 15]
```diff
@@ -0,0 +1,59 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
+    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
```
Comment (Teodor92) [inline tsconfig.json]: Lets switch to `es2017` we are running node 10 after all

File: tsconfig.json [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    // "preserveSymlinks": true,              /* Do not resolve the real path of symlinks. */
+
+    /* Source Map Options */
+    // "sourceRoot": "",                      /* Specify the location where debugger should locate TypeScript files instead of source locations. */
+    // "mapRoot": "",                         /* Specify the location where debugger should locate map files instead of generated locations. */
+    // "inlineSourceMap": true,               /* Emit a single file with source maps instead of having a separate file. */
+    // "inlineSources": true,                 /* Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set. */
+
+    /* Experimental Options */
+    // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
```
Comment (Teodor92) [inline tsconfig.json]: Lets activate this and the below one.

File: .gitignore [hunk 17]
```diff
@@ -59,3 +59,9 @@ typings/
 
 # next.js build output
 .next
+
+# output of the compilation
+dist/
+
+# package-lock.json file
+package-lock.json
```
Comment (Teodor92) [inline .gitignore]: Do not ignore the package-lock file - it is used for deterministic package restore :)

File: nodemon.json [hunk 18]
```diff
@@ -0,0 +1,8 @@
+{
+    "watch": ["src/**/*.ts"],
+    "exec": "ts-node src/index.ts",
+    "ignore": [
+        ".git",
+        "node_modules"
+      ]
+}
```
Comment (Teodor92) [inline nodemon.json]: A recommendation - always leave an empty line at the end of a file :) 

https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-file

Comment (Teodor92) [reply nodemon.json]: This is a comment for all the files. When we install linting - this will be automatically checked.

File: package.json [hunk 19]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "interview-me-backend",
+  "version": "1.0.0",
+  "description": "Question Quest is simple project which we decide to develop because of two main reasons. Be able to practice the following used technologies and finnaly have a place where we can store, operate and easily access interview questions and information",
```
Comment (Teodor92) [inline package.json]: What about:

Question Quest is project for practicing, storing, and easily accessing interview questions and information.

File: package.json [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+  "main": "index.ts",
+  "scripts": {
+    "start": "npm run build:watch",
+    "build": "tsc .",
```
Comment (Teodor92) [inline package.json]: Btw you don't even need to pass the root path as it is default. Just calling `tsc` will be enough :)

File: package.json [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    "build:watch": "nodemon"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/QuestionQuest/interview-me-backend.git"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
Comment (Teodor92) [inline package.json]: Btw the repo license is MIT not ISC

File: tsconfig.json [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+    "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
+    "noFallthroughCasesInSwitch": false,    /* Report errors for fallthrough cases in switch statement. */
+
+    /* Module Resolution Options */
+    "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
+    "esModuleInterop": true,                 /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
+    
+    /* Experimental Options */
+    "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
+
```
Comment (Teodor92) [inline tsconfig.json]: Empty line?

File: package.json [hunk 23]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "interview-me-backend",
+  "version": "1.0.0",
+  "description": "Question Quest is project for practicing, storing, and easily accessing interview questions and information.  ",
```
Comment (Teodor92) [inline package.json]: Do we need the empty space at the end of the string?

File: package.json [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+  "main": "index.ts",
+  "scripts": {
+    "start": "npm run build:watch",
```
Comment (Teodor92) [inline package.json]: Suggestion for the scripts. Lets have the following:
- start - build and runs the project ONCE. Does not watch for changes - useful when running in production.
- start:watch - build and runs the project continuously. Used when in dev.
- build - just builds the app - no running.

File: tsconfig.json [hunk 25]
```diff
@@ -0,0 +1,29 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
+    "target": "es2017",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
```
Comment (Teodor92) [inline tsconfig.json]: Nitpick - Can we remove all the side comments? We can get all that info from VSCode inteliscence.

**Comment by Teodor92** [comment]
One more thing - can we target `develop` with this PR?

File: package.json (hunk 23 above)
Comment (Okiana) [reply package.json]: no :)

File: package.json (hunk 24 above)
Comment (Okiana) [reply package.json]: Okay

**Comment by Teodor92** [comment]
You can merge this :)

**Pull request merged**